core/textile/backup.go

```diff
@@ -0,0 +1,266 @@
+package textile
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"github.com/FleekHQ/space-daemon/log"
+
+	"github.com/textileio/go-threads/core/thread"
+)
+
+// backup a single file with an io.Reader provided
+func (tc *textileClient) BackupFileWithReader(ctx context.Context, bucket Bucket, path string, reader io.Reader) (err error) {
+
+	if _, _, err = tc.UploadFileToHub(ctx, bucket, path, reader); err != nil {
+		log.Error(fmt.Sprintf("error backuping up file with reader (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	if _, err = tc.setMirrorFileBackup(ctx, path, bucket.Slug()); err != nil {
+		log.Error(fmt.Sprintf("error setting mirror file as backup (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	return nil
+}
+
+// backup a single file with a path provided
+func (tc *textileClient) backupFile(ctx context.Context, bucket Bucket, path string) error {
+
+	errc := make(chan error, 1)
+	pipeReader, pipeWriter := io.Pipe()
+
+	// go routine for piping
+	go func() {
```

Comment:
Why run this as a goroutine? It looks like you could has well just run it inline with the current execution stack and return any failed error and therefore won't be need for an error channel?

Reply:
@perfectmak 

This is a `io.Pipe` pattern. `bucket.GetFile()` writes to the `pipeWriter` and, at the same time, it is being read by `tc.BackupFileWithReader()`, so we don't need any temporary files. Please take a look at the [example](https://golang.org/pkg/io/#Pipe).

Reply:
Yes, I know of the pattern, but it still looks like it would still be executed sequentially expecially with the `if err := <-errc` used immediately after the go func that is supposed to write to the `pipeWriter`. And then after the err check you pass `pipeReader` into `tc.BackupFileWithReader()`.

If you have sometime, you should try testing it with the go routine and without and see if there is any difference.

Reply:
@perfectmak 

Great find. Thank you. I've tweaked this code so now it's truly parallel: https://github.com/FleekHQ/space-daemon/pull/191/files/593541272f2e1edb98679e8020351e5a6ecb5232#diff-b580f3c0539be5a72155b66b76313a90R53